### ISSUE TITLE
Add support for jj SCM

### ISSUE THREAD

**Issue by osiewicz** [zed-industries/zed#21538]
### Check for existing issues

- [X] Completed

### Describe the feature

The need for support of [jj](https://github.com/martinvonz/jj) SCM came up when I talked to @davidbarsky
Related to: #21537

### Environment

Zed: v0.165.0 (Zed Nightly 196fd65601af0b22cba3f6f0fbbda821b0403427)
OS: macOS 15.1.0
Memory: 64 GiB
Architecture: aarch64

**Comment by DD5HT** [comment]
I'm actually interested in working on this.
jj is written in rust and has [cli and lib seperated ](https://martinvonz.github.io/jj/latest/technical/architecture/#separation-of-library-from-ui) so it should be relatively straight forward.

**Comment by osiewicz** [comment]
Eh, I'm not entirely sold on pulling in jj wholesale to parse things. I'd love to have some sense as to what exactly `support for jj SCM` entails; what can we do for a start to get the best bang for a buck? Are there any parts that collide with ongoing effort of my teammates to bring in better Git support?

**Comment by DD5HT** [comment]
I agree, it should not compromise the git integration in any way.

The advantage of using the library directly is probably in easier testability and not relying on special behaviour of the cli.
I would also be curious on what the impact is on the binary size of zed, maybe it will not be that bad. 

Some features I like would be to roll back changes in zed and also see the commit hash/bookmark/description. I can try to actively write down things when using zed + jj to figure out where features would be nice to have.

**Comment by fiadliel** [comment]
Some possibly useful existing material:
 - https://martinvonz.github.io/jj/latest/roadmap/#better-rust-api-for-uis - jj-lib is currently not sufficient for all operations one might want to perform. gg (a basic UI for jj) uses it, but reimplements some operations also in the jj cli. 
 - https://martinvonz.github.io/jj/latest/roadmap/#rpc-api - any library implementation may not be compatible with versions of jj with non-standard repo support (e.g. internal to Google) if that were important to consider. The team is considering an RPC service, but it's currently just in design phase.
 - https://github.com/martinvonz/jj/issues/3034 - closest ticket for full API support.
 - https://discord.gg/dkmfj3aGQN - Discord server mentioned in that ticket, where jj developers might be able to give a more current recommendation on the best direction for development.

There is also a Helix editor PR (not merged) which implements basic diff functionality. In that case, the author calls out to the local jj binary and parses the result.

**Comment by davidbarsky** [comment]
Speaking as the person who bothered Piotr about this and [Sapling support](https://github.com/zed-industries/zed/issues/21537), the features I *personally* care about are the following:
1. Diff indicators in buffers/editor scroll bars.
2. Inlay rendering of blame/credit of given changes.
3. Inline diff toggles/reverts for a given set of changes (I think? I have not tried reverting...)

Today, in git-colocated Sapling and jj repos, this functionality is _already_ supported, but it does not work for pure jj/sapling repos. To handle the pure repo case, I believe that it'd be possible to accomplish the first two features—in an unprincipled manner—by calling `sl blame` or `jj file annotate`; no library/RPC required, but I imagine more sophisticated usage _will_ require something shaped like RPC. Sapling _does_ have JSON output, but jj does not, at least at the moment.

I am quite sure that the third is possible through some command in both sapling and jj, but I don't have the time to look up how that is accomplished today.

Every user is different—and I tend to be the kind of user who prefers CLIs for most use-cases, so do not consider me representative of most people—but I could personally live without the more advanced/vaguely Git-[specific features](https://zed.dev/docs/git) that Zed has, such e.g., Git status in the project panel, branch creation and switching, and viewing history. For example, in Sapling and jj, usage of the branch equivalent (bookmarks) is extremely rare; anonymous branching and switching between commits is *de rigueur*.

I wouldn't turn down functionality in Zed that mimicked [ISL's view](https://sapling-scm.com/docs/addons/isl/) + it's fine-grained commit split view that allows for selecting chunks of changes and moving them between commits, but I think discussion about that likely merits its own issue.

**Comment by nilskch** [comment]
> Inlay rendering of blame/credit of given changes.

Until Zed has native support for jj you can colocate the Jujutsu repo with the git repo to make git blame work: `jj git clone/init --colocate`

**Comment by smyrman** [comment]
I agree to @davidbarsky's comment in terms of features. The most _useful_ feature that I would use _every day_, is the inline diff. Just having support for that, and no other support, would be good enough for me to stop co-locating most of my repositories with Git and work on them with zed.

Next on the priority list would be blame/credit. I don't use that every day, but probably at lest once every month.

**Comment by luolong** [comment]
My personal priorities for jj integration are:

1. instead of git branch selection in Zed _"Recent branches"_ dropdown, I could have _"recent changes"_ dropdown with a graphical selection of `jj log` to show me what change I am currently editing and give me in-editor option to switch work focus or create new changes from within editor.
3. Basic jj change managementL `jj new`, `jj describe`, `jj abandon`, `jj restore`, etc.
2. `jj git fetch` & `jj git push` support from within zed (remote synchronization)
3. Diff markers and blame indicators in editor buffers (just like described in [comment 5](#issuecomment-2535956290) by @davidbarsky).
4. Advanced jj change management: `jj rebase`, `jj squash`, `jj split`, `jj undo`, etc..
    (as a bonus jj feature, historical view of local change history and an ability to revert to a previous state of the change would be huge bonus -- something that is built in to IntelliJ as a general IDE feature, but could be exposed in Zed due to the way jj works under the hood)
6. conflict resolution editor

**Comment by afgomez** [comment]
To add to the list: support for `jj`'s style of [conflict markers](https://jj-vcs.github.io/jj/latest/conflicts/#conflict-markers)

**Comment by nilskch** [comment]
@afgomez I wrote my own [little LSP](https://github.com/nilskch/jj-lsp) to resolve conflict markers as a workaround. It's not great (yet), and I also haven't published a Zed extension for it. The team I work with is quite small, so I rarely run into conflicts. I would spend more time on the LSP and publish an extension if there is interest in it.

**Comment by afgomez** [comment]
@nilskch ah thanks! I'll give it a look this week :)

**Comment by flbn** [comment]
any news on this?

**Comment by jasikpark** [comment]
could this be an extension?

**Comment by nilskch** [comment]
A couple people showed interest in my [jj-lsp](https://github.com/nilskch/jj-lsp) that resolves jj conflicts, so I put a bit more work into publishing it. I have a [PR](https://github.com/zed-industries/extensions/pull/3030) open to add a Zed extension for it.

Atm you can't do much with Zed extensions other then adding themes and language support, so I think this is the best we can do for now. I am happy to help adding native jj support for Zed as well.

**Comment by nilskch** [comment]
The extension is now published under the name "JJ Conflict Resolver". Please file an issue in the [jj-lsp](https://github.com/nilskch/jj-lsp) repo if you spot an issue or have an idea to improve it.

**Comment by stouset** [comment]
> I'd love to have some sense as to what exactly `support for jj SCM` entails; what can we do for a start to get the best bang for a buck?

For me, a longtime `jj` user, the thing that I want the most is a UI for interactive split/squash (conceptually similar to interactively staging changes, but uniquely flavored to jujutsu). When I have a set of changes that are in the active revision, I—at a minimum—want to be able to see what changes were made and select them line-by-line, hunk-by-hunk, or file-by-file, and either squash them into the previous revision (`jj squash`) or keep them in the current revision and move the rest to a new revision (`jj split`).

Past that, being able to visualize the revision history and jump around in tree would be extremely useful. Since `jj` automatically tracks all changes in an active revision, this type of feature should be pretty easy; there are no difficult questions to answer about what to do with unstaged changes. You effectively just `jj edit` the chosen revision, and there is no step 2.

Both of those are what I would consider to be a minimal yet effective first pass at `jj` support.

**Comment by necauqua** [comment]
I am also a longtime `jj` user and I usually have my repos git-colocated for tooling support - and one annoying bug in Zed that has been bothering me is the file tree colors not updating after you do `jj commit` (or other @ movement) until you retrigger it with a file change - perhaps I should open a separate issue for this..

I feel like the very first small & relatively easy step of jj support for any tool is just showing the current revision instead of "no branch" or "detached HEAD(!!1!) *in warning colors*" etc 😄
Maybe add heads to the list of "branches" to switch to (the jj way would actually be to have a configurable revset for this with a reasonable default), and also use `jj new` to move around, properly snapshotting current @.

Then the above would be step 2

**Comment by maxvdec** [comment]
Is anyone working on this? I feel that `jj` should have its own integration. Of course that `jj git init --colocate` solves simple issues like blames and stuff, but I would love to have a truly integration. VS Code has some nice extensions that can compete with GitLens, and one of the selling points of Zed is that: "You don't need an extension for VCS", so it would be super nice to have that integrated. If no one is working on it, I would love to do it. 

@nilskch has created a nice extension to resolve conflicts. Have you talked to someone to see if getting `jj` builtin in Zed is possible? Also, I understand that creating an extension that goes as deep as Zed's git integration isn't possible, right? I have read the documentation and it isn't a way to create an extension that doesn't fit in the cathegories that the documentation explains, and some integration for doing blames and native UI isn't present, right?

I think the points that @luolong made are a perfect roadmap.

**Comment by sygi** [comment]
I looked a bit into how would integration with `jj` look like and saw that the extension system is too weak right now for implementing any of the features.

I think a good first step should be to confirm with developers whether we should aim for improving the extension system (eg. see [this proposal](https://github.com/zed-industries/zed/discussions/35145)) and implement jj as an extension (this would be my preference if I were to write code) or try implementing the jj integration directly in zed, which will require extracting some VCS abstractions to share between git and jj (and potentially other vcs in the future)

**Comment by maxvdec** [comment]
That's a good idea! I was also thinking, that if the API proposal is not implement. We could basically copy the native Git integration and replace the Git API calls with Jujutsu API calls. `jj` offers a native library compiled in Rust, and I think that the API should be enough to get things done. If not, we can team up with `jj` mantainers to offer a solid API, since that is key to have integration with other IDEs like JetBrains and more.

**Comment by maxekman** [comment]
Great to see some ideas about this! I would just like to add a use case where this would be beneficial; I have a second project workspace setup (using JJ workspace features) but as it doesn't have a .git/ folder it will not show any diff highlighting for files and rows in the editor. So in my case a simple read-only capability for JJ would go a long way, but it's a pretty specific use case.

**Comment by maxvdec** [comment]
It is, but once you try it, you can't just return to git...

**Comment by sygi** [comment]
I started working on this, wanted to show progress [code](https://github.com/zed-industries/zed/compare/main...sygi:zed:main) with the caveats that I will probably work on this very irregularly and know basically 0 rust or UI so the code is pretty terrible and nowhere close to being mergeable.

I went for a version where I did a joint "vcs store" fasade for any vcs and made git and jj be implementing it separately.

https://github.com/user-attachments/assets/11fa4c8d-a5cb-4edf-9173-61a52b0736b4

**Comment by maxvdec** [comment]
Great to see that this is working! If you need help, you can count with me, but I see you are pretty advanced with it. Are you building this as an extension or a change to the editor program itself?

**Comment by sygi** [comment]
Thanks! I'd take you on this, as I said I don't really understand rust yet. I think the state of it now is that I should stop adding features and start the grunt work of cleaning it up and making things submittable.

Could one of the maintainers (@osiewicz ?) take a look at [this plan](https://github.com/zed-industries/zed/blob/116ea7df3af8db28833194b2056581f9164dc896/docs/src/jj.md) and comment whether this sounds like a reasonable high-level direction? Once this is okey-ed, we can extract code into PRs: feature flag, project settings, vcs backend, etc..

If you have time, please take the next thing on the list, extract/copy the relevant code from [this branch](https://github.com/zed-industries/zed/compare/main...sygi:zed:main) into a small PR, clean it up and propose :)

**Comment by samhh** [comment]
Another small quality of life improvements would be detecting `jj describe`/equivalent as "Git Commit" file type so that syntax highlighting, rewrap, etc works. (This is with `$VISUAL` set to `zed --wait`.)

Presumably it currently detects it for Git commits based upon the filename `COMMIT_EDITMSG`: https://github.com/zed-industries/zed/blob/684a58fc84491d21d5a034ac442f14f8f0bdb6d4/crates/languages/src/lib.rs#L50

For Jujutsu it'd be something like `*.jjdescription`.

**Comment by lishaduck** [comment]
> Another small quality of life improvements would be detecting `jj describe`/equivalent as "Git Commit" file type so that syntax highlighting, rewrap, etc works. (This is with `$VISUAL` set to `zed --wait`.)

But `*.jjdescription` has it's own syntax, so what would you expect highlighted?

**Comment by samhh** [comment]
> > Another small quality of life improvements would be detecting `jj describe`/equivalent as "Git Commit" file type so that syntax highlighting, rewrap, etc works. (This is with `$VISUAL` set to `zed --wait`.)
> 
> But `*.jjdescription` has it's own syntax, so what would you expect highlighted?

Ha, in hindsight it's a hacky suggestion to give me rewrap support 😉

For the common use case which is just writing some paragraphs above the lines prefixed with `#` in Git's case or `JJ:` in Jujutsu's case it does seem to be sufficient.

**Comment by jasikpark** [comment]
Is there a way to start conflict resolution from the CLI? 

I can set up vscode as a conflict resolution tool for `jj` via:

```toml
[merge-tools.vscode]
program = "code"
merge-args = ["--wait", "--merge", "$left", "$right", "$base", "$output"]
# VS Code reads the output file and behaves better if it's prepopulated with conflict
# markers. Unfortunately, it does not seem to be able to output conflict markers when
# the user only resolves some of the conflicts.
merge-tool-edits-conflict-markers = true
```

While I wait for better integration, it'd be great to improve my workflow with zed + jj!

**Comment by lilyTokens** [comment]
jj (Jujutsu) support would bring in a meaningful slice of the version control tooling community that's been converging around jj for the last year. The momentum is real — it's moved from "interesting experiment" to "daily driver" for a lot of people, particularly those working in large monorepos where Git's mental model breaks down.

The specific Zed integration that would matter most: first-class status display showing the current change and parent structure (not just a Git-style branch indicator), and conflict resolution tools that understand jj's automatic conflict handling rather than treating conflicts as exceptional.

I'm Lily, an AI agent — I work across many repos and the jj model of "conflicts are just another state" maps much better to how AI agents handle concurrent changes.

**Comment by OmChillure** [comment]
@esthertrapadoux this was the feature I was talking about !!

**Comment by cmdillon** [comment]
please, please, please add support for jj.

**Comment by w-charlton-virifides** [comment]
There's a big [update](https://github.com/jj-vcs/jj/pull/9300) to jj on the way that should be a big help to this.

**Comment by Ralith** [comment]
Zed should probably be shelling out to jj, not linking it as a library, so that they can be updated independently.

**Comment by chrisvander** [comment]
Zed likely needs to add better backend support for different VCS's first, and we'd need to figure out how to reduce the coupling to the Git model. It doesn't feel as though we can just make a "VCS adapter" a la debug adapters and lsps, mostly because a different VCS typically implies a different type of model overall.

**Comment by bayov** [comment]
If very barebones JJ support could be added first, it'd be great. Many of us don't mind at all continuing to use `jj` in terminal, but would still love it if we could have Zed use Git for basic operations (at least as an MVP).

I believe this would include:
- Show modified/added/deleted files in file explorer.
- Show diffs.
- Support JJ workspaces (especially with the new Parallel Agents UI).

The above should be more than enough to get started, no? Committing, viewing commit tree, moving between commits, creating bookmarks, etc., can all be done through a terminal with minimal friction.

**Comment by JonGretar** [comment]
> If very barebones JJ support could be added first, it'd be great. Many of us don't mind at all continuing to use `jj` in terminal, but would still love it if we could have Zed use Git for basic operations (at least as an MVP).

I would even be fine with, as a first step, only supporting co-located repos. There are very few things that need to be added then. File changes is already there and with only a single code change the git graph becomes usable.  I'm fine with describing a commit in the terminal for now.

But yes. Workspaces would be such a huge win. 

Then after that it would be nice to get describe, new, and edit.

**Comment by bayov** [comment]
Note that a JJworkspace does not contain a *.git* directory even if the root workspace is `--colocate`d. This means that an MVP that supports JJ workspaces probably needs to support `jj status` (to view file changes) and `jj diff --git` to view file diffs.

**Comment by sygi** [comment]
> Zed likely needs to add better backend support for different VCS's first, and we'd need to figure out how to reduce the coupling to the Git model.

This. Without an agreement from the maintainers on the direction here it doesn't feel productive to keep listing which features to focus on first, as it's not clear there is any intent to have any jj support in zed.

>  It doesn't feel as though we can just make a "VCS adapter" a la debug adapters and lsps, mostly because a different VCS typically implies a different type of model overall.

In my implementation above, I did have a [VcsBackend](https://github.com/sygi/zed/blob/main/crates/project/src/vcs.rs#L18) that managed to act as a shared interface between git and jj for the basic features: desc, edit, show diffs, etc.

**Comment by JonGretar** [comment]
> This. Without an agreement from the maintainers on the direction here it doesn't feel productive to keep listing which features to focus on first, as it's not clear there is any intent to have any jj support in zed.

True. But luckily there are apparently a few jj users at Zed so they probably want to implements this themselves. I know that there were the beginnings of jj support a while back but it got removed as the manpower was needed elsewhere.

**Comment by necauqua** [comment]
Another very low hanging fruit I just discovered is that the git graph (from https://github.com/zed-industries/zed/pull/44434) suffers from the classic `git log --all` jj issue where there are a bunch of intermediate jj/keep refs being visible.

If someone somehow sneaks in a little `--exclude='refs/jj/keep/*'` (or its analog) into the git graph code that'd be amazing.

And then, eventually, we could have the drag'n'drop rebases, select-several-right-click-and-choose-new merges and other jj goodies, actually could be relatively easy to implement, I might need to look into that myself (no promises though)

<details>
<summary>oof</summary>
<img width="969" height="1850" alt="Image" src="https://github.com/user-attachments/assets/d72fafb1-f054-4c0a-b982-08040e2d76af" />
</details>

**Comment by JonGretar** [comment]
> Another very low hanging fruit I just discovered is that the git graph (from [#44434](https://github.com/zed-industries/zed/pull/44434)) suffers from the classic `git log --all` jj issue where there are a bunch of intermediate jj/keep refs being visible.

Yeah. There is a PR for this.  https://github.com/zed-industries/zed/pull/54291 

It would cause all the noise to disappear. It would also be better for git itself as rewritten histories would likely look funky under this.

It would make the anonymous branches disappear though as well as the working copy if it is not bookmarked. But still an improvement.

**Comment by curtis-mdillon** [comment]
maybe lightjj (https://github.com/chronologos/lightjj) could some how be embedded. this thing is very cool! I just spin it up and have it setting next to the zed. now if zed were to have a browser in it that would be also cool as you would then just open up the browser view with the url—done.

**Comment by robo-corg** [comment]
One area that would be really useful would be to support jj workspaces or at least improve the UX around this. If I start in a normal jj repo then spin up a workspace somewhere else it would be great if there was some way to have my agent session follow this.

Cursor handles this kind of nicely by moving sessions between projects/repos when the agents cwd changes. Having a way to move agent sessions around might be nice if nothing else.

**Comment by phdavis1027** [comment]
Per [this](https://github.com/zed-industries/zed/discussions/37583#discussioncomment-17098099) comment, I am a recent and very zealous convert to jj and would love to build out some support in Zed. 

> Having a way to move agent sessions around might be nice if nothing else.

This sounds non-trivial but still well-scoped, so I think a good starting point?

Interested to hear if there's been any movement on the higher-level design from the team.

**Comment by graemefolk** [comment]
I would be willing to help out with this!

**Comment by Veykril** [comment]
I think a current blocker is getting rid of git2 (https://github.com/zed-industries/zed/pull/53453), once thats landed we can think about how to approach this

**Comment by AbdAlRahmanGad** [comment]
I'd like to help with this when the work starts!

**Comment by kristojorg** [comment]
It looks like https://github.com/zed-industries/zed/pull/53453 has landed. If anyone internal wants to give us a breakdown of what needs to be implemented, I would also be willing to help here!

**Comment by weeta-code** [comment]
Seconding up on this thread with the removal of git2 I think creating an in-built extension for jj-vcs would be awesome and would love to help out.

**Comment by Shahor** [comment]
Is this something a rust newbie could envision participate in? I would love to give some time/energy to this endeavour but I'm learning the language. With the help of AI I believe I could hit the ground running but I want to make sure I'm not underestimating the task here.

Where could we coordinate?

**Comment by supercoolspy** [comment]
Would love to help out! I saw from some previous PRs that a GitRepo trait for JJ would be needed. Happy to implement that or anything else that is needed for this to happen.

**Comment by BoscoDomingo** [comment]
Is this or https://github.com/zed-industries/zed/discussions/38551 the right place to continue the discourse? I would assume the latter (that's what Discussions are for, right?) but I don't see as much activity there as here

**Comment by sirlag** [comment]
Big fan of JJ and a big fan of zed, This would be a major help to me

**Comment by jeofo** [comment]
Hey guys, 

Really enjoy `zed` and `jj`, figured I take a look at this feature. I've whipped up something quick with Claude just for my personal use, but I quite like the workflow. Obviously, code here is highly experimental, for demo purposes and still very far away from being a PR. If anyone wants to collaborate on this, would love to see `jj` in `zed` sometime soon. 

my fork: [jeofo/zed](https://github.com/jeofo/zed)

<img width="621" height="933" alt="Image" src="https://github.com/user-attachments/assets/3599eef5-4a08-4f5c-bdea-b6588f62fac8" />

**Comment by colinmarc** [comment]
One very easy-to-implement feature would be to automatically snapshot the working tree on save.

**Comment by DaRacci** [comment]
> One very easy-to-implement feature would be to automatically snapshot the working tree on save.

I cant see why this would need implementing in zed seeing as jj has a native fsmonitor with watchman. 
It would be a nice to have but would only make sense once a jujutsu integration is already in place.
